Add working example to CSS @property page (#27688)
mdn/content: files/en-us/web/css/@property/index.md

{{CSSRef}}

> The **`@property`** [CSS](/en-US/docs/Web/CSS) [at-rule](/en-US/docs/Web/CSS/At-rule) is part of the [CSS Houdini](/en-US/docs/Web/Guide/Houdini) umbrella of APIs. It allows developers to explicitly define their {{cssxref('--*', 'CSS custom properties')}}, allowing for property type checking and constraining, setting default values, and defining whether a custom property can inherit values or not.

Reviewer: note that i edited this paragraph, so please review my changes ;)

{{ EmbedLiveSample('examples', '100%', '250px') }}

The two custom properties,  `--item-size: 20%` and `--item-color: orange;` are set on the `container` parent, overriding the `40%` and `aqua` default values set when these custom properties were defined. The size is set to be inheritable; the color is not.

For item one, none of these custom properties have been set. The `--item-size` is inheritable, so the value `20%` set on its parent `container` is used. On the other hand, the property `--item-color` is not inheritable, so the value `orange` set on the parent is not considered. Instead the default initial value `aqua` is  used.

For item two, CSS global keywords are set for both custom properties which are valid values for all value types and therefore valid no matter the `syntax` descriptor value. The `--item-size` is set to `initial` and uses the `initial-value: 40%;` set in the `@property` declaration.  The `initial` value means the`initialValue` value for the property is used. The `--item-color` is set to `inherit`, explicitly inheriting the `orange` value from its parent even though the custom property is set to otherwise not be inherited. This is why item two is orange.

> For item three, the `--item-size` value gets set to `1000px`. While `1000px` is a {{cssxref("length") value, the `@property` declaration require the value be a `<percentage>`, so the declaration is not valid and is ignored, meaning the inheritable `20%` set on the parent is used. The `xyz` value is also invalid. As `registerProperty()` set `--item-color` to not be inherited, default initial value of `aqua` is used and not the parent's `orange` value.

Reviewer: ```suggestion
For item three, the `--item-size` value gets set to `1000px`. While `1000px` is a {{cssxref("length") value, the `@property` declaration requires the value be a `<percentage>`, so the declaration is not valid and is ignored, meaning the inheritable `20%` set on the parent is used. The `xyz` value is also invalid. As `registerProperty()` set `--item-color` to not be inherited, the default initial value of `aqua` is used and not the parent's `orange` value.
```

```js
window.CSS.registerProperty({
  name: "--item-color",
  syntax: "<color>",
  inherits: false,
  initialValue: "aqua",
});
```

Following code creates the items that we want to style using our custom properties.

```html
<div class="container">
  <div class="item one">Item one</div>
  <div class="item two">Item two</div>
  <div class="item three">Item three</div>
</div>
```

Now let us style the items.

```css
.container {
  display: flex;
  height: 200px;
  border: 1px dashed black;

  /* set custom property values on parent */
  --item-size: 20%;
  --item-color: orange;
}

/* use custom properties to set item size and background color */
.item {
  width: var(--item-size);
  height: var(--item-size);
  background-color: var(--item-color);
}

/* set custom property values on element itself */
.two {
  --item-size: initial;
  --item-color: inherit;
}

.three {
  /* invalid values */
  --item-size: 1000px;
  --item-color: xyz;
}
```

{{ EmbedLiveSample('examples', '100%', '250px') }}

> In the output above, for the property `--item-size`:
> 
> - for item one, as we have not set any value in any CSS rule for the item, value `20%` is inherited from the parent `.container`.
> - for item two, the `initial` value means to use the default `40%` from the property definition.
> - for item three, value `1000px` is not a valid value, because it is not a percentage. So the inherited value `20%` is used.
> 
> In the output above, for the property `--item-color`:
> 
> - for item one, as the property is not inheritable and it is not set in any CSS rule for the item, default value `aqua` is used.
> - for item two, using `inherit` we've forced the inheritance, even though the property is not inheritable, the value `orange` set on parent is used.
> - for item three, value `xyz` is an invalid color, so default `aqua` is used.

Reviewer: There are issues with content here. You may want to change your example, or go with an improved version of the following, which is what is actually happening:

```suggestion
The two custom properties,  `--item-size` and `--item-color: orange;` are set on the `container` parent. As no values are reset for item one, the `20%` and orange are inherited from the parent.

Both custom properties have values set in item two, but one value does not conform to the syntax requirements set in property registration. The `registerProperty()` method requires a `<color>` value for `--item-color`. While `inherit` is a valid value for `background-color`, it is not a valid `<color>` value as required by the property registration, so the declaration is ignored. The `background-color` declaration from `.item` is valid and is therefore used, setting the background color to `orange`. (Had `background-color` been inherited, as the container has no background color, item two would also have had no background color.) The `--item-size` custom property is set `initial`. This property was not registered, and therefore there is no syntax requirement. The `initial` value means the default `content` value for the property is used.

For item three, the `--item-size` value gets set to `1000px`. This is valid. As `xyz` is not a valid CSS keyword, the initial value of `aqua`, set in the property registration, is used.
```

Author: Unfortunately, the live sample in preview page is not showing the output.\
Please look at exact replication of this here: https://jsfiddle.net/hcbuL48a/
I am afraid the output doesn't conform with your comments.
 
> As no values are reset for item one, the `20%` and orange are inherited from the parent.

On line 63, in JavaScript, we've made `--item-color` non inheritable so the item `.one` _doesn't_ inherit orange color. In the out put it stays the default value `aqua`.

> For item three, the `--item-size` value gets set to `1000px`. This is valid.

No as per the declaration on line 51 the `--item-size` accepts only percentage values. So `1000px` is not a valid value. In the output the item gets the inherited `20%` size.

I've tried to demonstrate inheritable(`--item-size`) as well as non inheritable(`--item-color`) properties in the example. I'll split it once we agree on the current output.

Reviewer: I missed the CSS 

```css
 @property --item-size {
  syntax: "<percentage>";
  inherits: true;
  initial-value: 40%;
}
```
my review is incorrect. i have to rereview. Apologies. I will need to readress this review

Author: The explanation for `one` is put at the end of the second para for `two`. Mind keeps wondering what happened to `one`.
How about
```suggestion
The two custom properties,  `--item-size: 20%` and `--item-color: orange;` are set on the `container` parent, overriding the `40%` and `aqua` default values set when these custom properties were defined. The size is set to be inheritable; the color is not.

For item one, none of these custom properties have been set. The `--item-size` is inheritable, so the value `20%` set on its parent `container` is used. On the other hand, the property `--item-color` is not inheritable, so the value `orange` set on the parent is not considered. Instead the default initial value `aqua` is  used.

For Item two, CSS global keywords are set for both custom properties which are valid values for all value types and therefore valid no matter the `syntax` descriptor value. The `--item-size` is set to `initial` and uses the `initial-value: 40%;` set in the `@property` declaration.  The `initial` value means the`initialValue` value for the property is used. The `--item-color` is set to `inherit`, explicitly inheriting the `orange` value from its parent even though the custom property is set to otherwise not be inherited. This is why item two is orange.

For item three, the `--item-size` value gets set to `1000px`. While `1000px` is a {{cssxref("length") value, the `@property` declaration require the value be a `<percentage>`, so the declaration is not valid and is ignored, meaning the inheritable `20%` set on the parent is used. The `xyz` value is also invalid. As `registerProperty()` set `--item-color` to not be inherited, default initial value of `aqua` is used and not the parent's `orange` value.
```

Unknown descriptors are invalid and ignored, but do not invalidate the `@property` rule.

## Formal syntax

{{csssyntax}}

## Examples

In this example, we define two custom properties, `--item-size` and `--item-color`, that we'll use to define the size (width and height) and background color of the three following items.

```html
<div class="container">
  <div class="item one">Item one</div>
  <div class="item two">Item two</div>
  <div class="item three">Item three</div>
</div>

> \```

Author: ```suggestion
```
```

Author: **Tip:** If we we want to show backquotes(`` ` ``) in a code block, then we need to increase number of backquotes in the fence by one.\
For example, to show two backquotes(``` `` ```) in a code block we need to use three backqoutes in fence:
```md
For example, to show two backquotes(``` `` ```) in a code block we need to use three backqoutes in fence:
```

For PR suggestion containing backquotes increase the number of backquotes in the ```` ```suggestion ```` fence:
`````
````suggestion
```html
<!-- some code -->
```
````
`````
To show this code block in this GitHub comment, it required 5 backquotes :p
